homeassistant/components/mikrotik/__init__.py

```diff
@@ -1 +1,390 @@
+                         "local wireless interfaces.", (host))
+            capsman_exist = False
+        try:
+            wireless_exist = self._client[host](
+                cmd=MIKROTIK_SERVICES[WIRELESS])
+        except (librouteros.exceptions.TrapError,
+                librouteros.exceptions.MultiTrapError,
+                librouteros.exceptions.ConnectionError):
+            wireless_exist = False
+
+        if not wireless_exist and not capsman_exist \
+                or polling_method == 'ip':
+            _LOGGER.info(
+                "Mikrotik %s: Wireless adapters not found. Try to "
+                "use DHCP lease table as presence tracker source. "
+                "Please decrease lease time as much as possible",
+                host)
+        if polling_method:
+            _LOGGER.info("Mikrotik %s: Manually selected polling method %s",
+                         host, polling_method)
+            device_tracker = polling_method
+        else:
+            if capsman_exist:
+                device_tracker = CAPSMAN
+            elif wireless_exist:
+                device_tracker = WIRELESS
+            else:
+                device_tracker = DHCP
+            _LOGGER.info("Mikrotik %s: Using device_tracker method %s",
+                         host, device_tracker)
+        self._hosts[host][DEVICE_TRACKER] = device_tracker
+
+    def connect_to_device(self, host):
+        """Connect to Mikrotik method."""
+        import librouteros
+        import time
+        tick = 0
+        while self._hosts[host][CONNECTING] and tick < 60:
+            time.sleep(1)
+            tick += 1
+        if self._hosts[host][CONNECTED]:
+            return
+        self._hosts[host][CONNECTING] = True
+        self._hosts[host][CONNECTED] = False
+        _LOGGER.debug("[%s] Connecting to Mikrotik device.", host)
+        config = self._hosts[host]['config']
+        self.config_kwargs(config)
+        try:
+            self._client[host] = librouteros.connect(
+                config[CONF_HOST], config[CONF_USERNAME],
+                config.get(CONF_PASSWORD, ''),
+                **self._hosts[host]['kwargs'])
+        except (librouteros.exceptions.TrapError,
+                librouteros.exceptions.MultiTrapError,
+                librouteros.exceptions.ConnectionError) as api_error:
+            _LOGGER.error(
+                "Mikrotik error for device %s. "
+                "Connection error: %s", host, api_error)
+            self._hosts[host][CONNECTING] = False
+            self._hosts[host][CONNECTED] = False
+            self._client[host] = None
+            return False
+
+        host_name = (self._client[host](
+            cmd=MIKROTIK_SERVICES[IDENTITY]))[0]['name']
+        if not host_name:
+            _LOGGER.error("Mikrotik failed to connect to %s.", host)
+            return False
+        self.hass.data[MIKROTIK][host]['name'] = host_name
```

Comment:
It looks like we are storing information from the clients in different dictionaries, while instead we should group this information as attributes of the clients or make a new client class that holds the api client and any other information we need to store. Then we just need to store all the client instances in one dictionary.

Reply:
I will rework this code into using a MikrotikClient class and store the api client like you suggested. Thanks.

Reply:
I committed new code using a MikrotikClient class. I didn't see CI run code validation on it. How can I check if there are any errors/issues?